src/vision/status_quo/alan_builds_a_task_scheduler.md

```diff
@@ -0,0 +1,698 @@
+			fmt.Fprintf(w, "failed to read request %v", err)
+			return
+		}
+		s, err := applier.apply(Request{echoRequest: &echoRequest})
+		if err != nil {
+			fmt.Fprintf(w, "failed to apply request %v", err)
+			return
+		}
+		fmt.Fprint(w, s)
+
+	default:
+		http.Error(w, "Method Not Allowed", 405)
+	}
+}
+```
+
+For Rust, Alan has multiple options to build a web server: [hyper](https://github.com/hyperium/hyper), [actix-web](https://github.com/actix/actix-web), [warp](https://github.com/seanmonstar/warp), [rocket](https://github.com/SergioBenitez/Rocket), [tide](https://github.com/http-rs/tide), etc..
+
+Alan strongly believes in Go's minimal dependency approach, and thereby chooses "hyper" for its low-level API. While "hyper" is meant to be a low-level building block, implementing a simple request handler in "hyper" still requires four different external dependencies. Alan is not surprised anymore, and rather accepts the status quo of split Rust ecosystem:
+
+```bash
+cargo add http
+cargo add futures
+cargo add hyper --features full
+cargo add tokio --features full
+```
+
+After multiple days, Alan finally writes the following code:
+
+```rust
+use async_std::sync::Arc;
+use futures::TryStreamExt;
+use http::{Method, Request, Response, StatusCode, Version};
+use hyper::server::conn::AddrStream;
+use hyper::service::{make_service_fn, service_fn};
+use hyper::{Body, Server};
+use tokio::signal;
+
+pub struct Handler {
+    listener_port: u16,
+    applier: Arc<apply::Applier>,
+}
+
+impl Handler {
+    ...
+    pub async fn start(&self) -> Result<(), Box<dyn std::error::Error>> {
+        println!("starting server");
+        match self.applier.start().await {
+            Ok(_) => println!("started applier"),
+            Err(e) => panic!("failed to stop applier {}", e),
+        }
+
+        let addr = ([0, 0, 0, 0], self.listener_port).into();
+        let svc = make_service_fn(|socket: &AddrStream| {
+            let remote_addr = socket.remote_addr();
+            let applier = self.applier.clone();
+            async move {
+                Ok::<_, Infallible>(service_fn(move |req: Request<Body>| {
+                    handle_request(remote_addr, req, applier.clone())
+                }))
+            }
+        });
+
+        let server = Server::bind(&addr)
+            .serve(svc)
+            .with_graceful_shutdown(handle_sigint());
+
+        if let Err(e) = server.await {
+            println!("server error: {}", e);
+        }
+
+        match self.applier.stop().await {
+            Ok(_) => println!("stopped applier"),
+            Err(e) => println!("failed to stop applier {}", e),
+        }
+
+        Ok(())
+    }
+}
+
+async fn handle_request(
+    addr: SocketAddr,
+    req: Request<Body>,
+    applier: Arc<apply::Applier>,
+) -> Result<Response<Body>, hyper::Error> {
+    let http_version = req.version();
+    let method = req.method().clone();
+    let cloned_uri = req.uri().clone();
+    let path = cloned_uri.path();
```

Comment:
It would be better if we don't need to clone this. We can just do `req.uri().path()` below.

Side tip, people will usually try to avoid `clone()` and `unwrap()` in rust, since it's not always necessary, but sometimes it will make life easier.